### ISSUE TITLE
bps.push_pull give wrong result for pytorch

### ISSUE THREAD

**Issue by vycezhong** [bytedance/byteps#272]
When I train mnist with pytorch, I found the output accuracy and loss are werid. Then I tried to print it out before push_pull. 

![Screen Shot 2020-07-18 at 8 24 08 PM](https://user-images.githubusercontent.com/25879526/87852520-381b8a00-c935-11ea-9293-c1f0df956a50.png)

So I guess this is because `push_pull` gives wrong results for CPU tensors. 
https://github.com/bytedance/byteps/blob/cf020c97fc718ca209cbadbfac4cffa5e49d7d21/example/pytorch/train_mnist_byteps.py#L133

Actually it is a known bug which was also found in MXNet I have mentioned before. #247 
Current workaround is to set the tensor to cuda. 

```py
tensor = torch.tensor(val).cuda()
```

**Comment by bobzhuyb** [comment]
We also found that push pull on CPU tensor may lead to instability. We will check and fix it.
